dask_expr/expr.py

```diff
@@ -206,8 +132,35 @@ def _layer(self) -> dict:
         Expr._task
         Expr.__dask_graph__
         """
+        raise NotImplementedError(
+            "Expressions should define either _layer (full dictionary) or _task"
+            " (single task).  This expression type defines neither"
+        )
 
-        return {(self._name, i): self._task(i) for i in range(self.npartitions)}
+    def _depth(self):
+        """Depth of the expression tree
+
+        Returns
+        -------
+        depth: int
+        """
+        if not self.dependencies():
+            return 1
+        else:
+            return max(expr._depth() for expr in self.dependencies()) + 1
+
+    def operand(self, key):
+        # Access an operand unambiguously
+        # (e.g. if the key is reserved by a method/property)
+        return self.operands[type(self)._parameters.index(key)]
+
+    def dependencies(self):
+        # Dependencies are `Expr` operands only
+        return [operand for operand in self.operands if isinstance(operand, Expr)]
+
+    @property
+    def npartitions(self):
+        raise NotImplementedError
```

Comment:
Maybe npartitions should move over?  Not a huge deal I guess if collections like scalar or delayed return 1 though.

Reply:
I checked this with the dask/dask implementation and for stuff like bag this is given through __init__, that's why I ordered it like this